### PR TITLE
tidb: support add location labels

### LINKED ISSUE #4662: Also add location labels to tidb
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by glorv**
### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [x] Unit test 
- [ ] E2E test 
- [x] Manual test 
  - I uses kind to deploy a cluster in my local environment and edit the node `kind-control-plane` with a new label `topology.kubernetes.io/zone: test`. Then I add a pd config `location-labels = ["topology.kubernetes.io/zone", "kubernetes.io/hostname"]` to "example/basic/tidb-cluster.yaml" and deploy it to set up a cluster. When set the cluster version with 6.1.0, there is no logs about set tidb lables as expected. Then I changed the min version to 6.1.0 and redeploy operator, tidb log show logs like `[2022/08/10 09:43:19.460 +00:00] [INFO] [http_handler.go:2186] ["update server labels"] [labels="{\"kubernetes.io/hostname\":\"kind-control-plane\",\"topology.kubernetes.io/zone\":\"test\",\"zone\":\"test\"}"]` That show labels had been set to tidb successfully.
- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
None
```

### CODE REVIEW

**Comment by glorv** [comment]
@DanielZhangQD  @handlerww PTAL

File: pkg/controller/tidb_control.go [hunk 1]
```diff
@@ -137,6 +140,23 @@ func (c *defaultTiDBControl) GetSettings(tc *v1alpha1.TidbCluster, ordinal int32
...
+func (c *defaultTiDBControl) SetServerLabels(tc *v1alpha1.TidbCluster, ordinal int32, labels map[string]string) error {
+	httpClient, err := c.getHTTPClient(tc)
+	if err != nil {
+		return err
+	}
+
+	buffer := bytes.NewBuffer(nil)
+	if err := json.NewEncoder(buffer).Encode(labels); err != nil {
+		return fmt.Errorf("encode labels to json failed, error: %v", err)
+	}
```
Comment (KanShiori) [inline pkg/controller/tidb_control.go:153]: just confuse, why not use `json.Marshal`?

Comment (KanShiori) [reply pkg/controller/tidb_control.go:153]: ignore, I got

File: pkg/controller/tidb_control.go [hunk 2]
```diff
@@ -137,6 +140,23 @@ func (c *defaultTiDBControl) GetSettings(tc *v1alpha1.TidbCluster, ordinal int32
...
+
+	url := fmt.Sprintf("%s/labels", c.getBaseURL(tc, ordinal))
```
Comment (nolouch) [inline pkg/controller/tidb_control.go:155]: Does it works?

Comment (glorv) [reply pkg/controller/tidb_control.go:155]: The PR in tidb side https://github.com/pingcap/tidb/pull/36845 is merged.

**Comment by glorv** [comment]
@KanShiori @handlerww  PTAL

File: pkg/manager/member/tidb_member_manager.go [hunk 3]
```diff
@@ -1044,6 +1053,86 @@ func (m *tidbMemberManager) syncTidbClusterStatus(tc *v1alpha1.TidbCluster, set
...
+			klog.Warningf("node: [%s] has no node labels, skipping set store labels for Pod: [%s/%s]", nodeName, ns, name)
+			continue
+		}
+		// add the special `zone` label because tidb depends on this label for follower read.
+		labels[tidbDCLabel] = labels[zoneLabel]
+
+		if err := m.deps.TiDBControl.SetServerLabels(tc, ordinal, labels); err != nil {
+			klog.Warningf("cluster %s/%s set server labels for pod %s failed, error: %v", ns, tc.GetName(), name, err)
+			continue
+		}
```
Comment (KanShiori) [inline pkg/manager/member/tidb_member_manager.go:1111]: If tidb version is old, does it always log warnnings?

File: pkg/manager/member/tidb_member_manager.go [hunk 4]
```diff
@@ -1044,6 +1053,86 @@ func (m *tidbMemberManager) syncTidbClusterStatus(tc *v1alpha1.TidbCluster, set
...
+		if err != nil {
+			return setCount, err
+		}
+
+		pod, err := m.deps.PodLister.Pods(ns).Get(name)
+		if err != nil {
+			return setCount, fmt.Errorf("setServerLabels: failed to get pods %s for cluster %s/%s, error: %s", name, ns, tc.GetName(), err)
+		}
+
+		nodeName := pod.Spec.NodeName
```
Comment (KanShiori) [inline pkg/manager/member/tidb_member_manager.go:1099]: How about using the node name in tidb status, see L1042

File: pkg/manager/member/tidb_member_manager.go (hunk 3 above)
Comment (glorv) [reply pkg/manager/member/tidb_member_manager.go:1118]: I have added a min tidb version check at the beginning of this method

**Comment by glorv** [comment]
@KanShiori PTAL again, thank you

File: pkg/manager/member/tidb_member_manager.go (hunk 3 above)
Comment (KanShiori) [reply pkg/manager/member/tidb_member_manager.go:1118]: maybe use the `pkg/util/cmpver` pkg to compare version, it works for dirty versions or the latest version.

Comment (KanShiori) [reply pkg/manager/member/tidb_member_manager.go:1118]: For example:
https://github.com/pingcap/tidb-operator/blob/00002ede5e5b2c811d8bbff4043a4cc7a2a1e2a6/pkg/manager/member/tiflash_util.go#L109-L115

**Comment by KanShiori** [comment]
@glorv I left some advice about version check.

BTW, please test this feature and add the result in the `Manual test` section.

File: pkg/manager/member/tidb_member_manager.go (hunk 3 above)
Comment (glorv) [reply pkg/manager/member/tidb_member_manager.go:1119]: updated

**Comment by glorv** [comment]
> @glorv I left some advice about version check.
> 
> BTW, please test this feature and add the result in the `Manual test` section.

I have update the version change and manual test result. PTAL again thanks @KanShiori

File: pkg/manager/member/tidb_member_manager.go [hunk 5]
```diff
@@ -1044,6 +1054,93 @@ func (m *tidbMemberManager) syncTidbClusterStatus(tc *v1alpha1.TidbCluster, set
 	return nil
 }
 
+const tidbSupportLabelsMinVersin = "6.2.0"
+
+func (m *tidbMemberManager) setServerLabels(tc *v1alpha1.TidbCluster) (int, error) {
+	if tc.Spec.TiDB.Version != nil && (*tc.Spec.TiDB.Version) != "" {
```
Comment (KanShiori) [inline pkg/manager/member/tidb_member_manager.go:1060]: If the client use the `image` field not `version`, the version is empty. How about adding a function `TiDBVersion()` to parse version, like `TiFlashVersion`.
```go
// TiFlashVersion returns the image version used by TiFlash.
//
// If TiFlash isn't specified, return empty string.
func (tc *TidbCluster) TiFlashVersion() string {
	if tc.Spec.TiFlash == nil {
		return ""
	}

	image := tc.TiFlashImage()
	colonIdx := strings.LastIndexByte(image, ':')
	if colonIdx >= 0 {
		return image[colonIdx+1:]
	}

	return "latest"
}
```

In the future, we just need modify this function when we remove the `image` filed.

Comment (glorv) [reply pkg/manager/member/tidb_member_manager.go]: done.

**Comment by glorv** [comment]
@DanielZhangQD  @handlerww  PTAL

**Review by handlerww — APPROVED** [review]
lgtm

**Comment by glorv** [comment]
@KanShiori Shall we merge this PR now?

**Comment by glorv** [comment]
@KanShiori  Shall we merge this PR now?

**Comment by kolbe** [comment]
@glorv why doesn't this PR include release notes?

**Pull request merged**